saltpack.go

```diff
@@ -0,0 +1,135 @@
+// Copyright 2015 Keybase, Inc. All rights reserved. Use of
+// this source code is governed by the included BSD license.
+
+package updater
+
+import (
+	"encoding/hex"
+	"fmt"
+	"io"
+
+	"github.com/agl/ed25519"
+	"github.com/keybase/go-logging"
+	"github.com/keybase/saltpack"
+)
+
+type naclSigningKeyPublic [ed25519.PublicKeySize]byte
+type naclSignature [ed25519.SignatureSize]byte
+
+const kidNaclEddsa = 0x20
+
+// KID is a key identifier
+type KID string
+
+const (
+	// KIDLen is KID length in bytes
+	KIDLen = 35
+	// KIDSuffix is KID suffix (byte)
+	KIDSuffix = 0x0a
+	// KIDVersion is current version of KID
+	KIDVersion = 0x1
+)
+
+// KIDFromRawKey returns KID from bytes by type
+func KIDFromRawKey(b []byte, keyType byte) KID {
+	tmp := []byte{KIDVersion, keyType}
+	tmp = append(tmp, b...)
+	tmp = append(tmp, byte(KIDSuffix))
+	return KIDFromSlice(tmp)
+}
+
+// KIDFromSlice returns KID from bytes
+func KIDFromSlice(b []byte) KID {
+	return KID(hex.EncodeToString(b))
+}
+
+// Equal returns true if KID's are equal
+func (k KID) Equal(v KID) bool {
+	return k == v
+}
+
+// SaltpackVerifyDetached verifies a message signature
+func SaltpackVerifyDetached(reader io.Reader, signature string, validKIDs []KID, log logging.Logger) error {
+	if reader == nil {
+		return fmt.Errorf("Saltpack Error: No reader")
+	}
+	checkSender := func(key saltpack.SigningPublicKey) error {
+		kid := SigningPublicKeyToKeybaseKID(key)
+		log.Infof("Signed by %s", kid)
+		if kidIsIn(kid, validKIDs) {
+			log.Debug("Valid KID")
+			return nil
+		}
+		return fmt.Errorf("Unknown signer KID: %s", kid)
+	}
+	return SaltpackVerifyDetachedCheckSender(reader, []byte(signature), checkSender)
+}
+
+func kidIsIn(k KID, list []KID) bool {
+	for _, h := range list {
+		if h.Equal(k) {
+			return true
+		}
+	}
+	return false
+}
```

Comment:
Maybe pass the kids in a `map[KID]struct{}` instead of a slice, and avoid needing to write this function?